### PR TITLE
DefaultshareandRbac Check

### PR DESCRIPTION
**Pull request by MainiSuruchi**

### CODE REVIEW

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 1]
```diff
@@ -4006,11 +4007,102 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                    -RequiredModules @("Microsoft.Graph.Users", "Microsoft.Graph.Groups", "Microsoft.Graph.Identity.DirectoryManagement")
+                    
+            $userOid = $(Get-MgUser -Filter "UserPrincipalName eq $UserPrincipalName" -Property Id).Id
+
+            $groupIds = Get-MgUserMemberOf -UserId $userOid | Select-Id
+
+            $roleNames = @(
+            "Storage File Data SMB Share Reader",
+            "Storage File Data SMB Share Contributor",
+            "Storage File Data SMB Share Elevated Contributor"
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4069]: Nit: indent these

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 2]
```diff
@@ -4006,11 +4007,102 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+            )
+            $listOfRoleNames = @{}
+            foreach ($roleName in $roleNames) {
+                $assignments = Get-AzRoleAssignment -RoleDefinitionName $roleName
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4073]: We need to also look at the _scope_ of the role assignment. It must be a role assignment that applies to this storage account, and not to other storage accounts. Or alternatively, it should apply to the resource group or subscription, and be inherited.

**Review by MaximeKjaer — CHANGES_REQUESTED** [review]

File: AzFilesHybrid/AzFilesHybrid.psm1 (hunk 2 above)
Comment (MaximeKjaer) [reply AzFilesHybrid/AzFilesHybrid.psm1:4073]: Did some experimentation, something like this should work:

```powershell
$scope = "/subscriptions/$subscription/resourceGroups/$resourceGroup/providers/Microsoft.Storage/storageAccounts/$storageAccount/fileServices/default/fileshares/$fileShare"
$assignments = Get-AzRoleAssignment -RoleDefinitionName $roleName -Scope $scope
```

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 3]
```diff
@@ -3790,7 +3787,7 @@ function Debug-AzStorageAccountEntraKerbAuth {
             "CheckRegKey" = [CheckResult]::new("CheckRegKey");
             "CheckKerbRealmMapping" = [CheckResult]::new("CheckKerbRealmMapping");
             "CheckAdminConsent" = [CheckResult]::new("CheckAdminConsent");
-            ""=[]::new("") 
+            "CheckDefaultShareRBACPermission"=[CheckResult]::new("CheckDefaultShareRBACPermission")
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:3790]: What would you think of simplifying this name to something like "CheckRbac" ?

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 4]
```diff
@@ -4019,17 +4016,26 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                    Write-Error "CheckDefaultShareRBACPermission - FAILED"
+                }
+                else 
+                {
+                    $DefaultSharePermission = $StorageAccountObject.AzureFilesIdentityBasedAuth.DefaultSharePermission
+                    
+                    if((!$DefaultSharePermission) -or ($DefaultSharePermission -eq 'None'))
+                    {
+                        Debug-RBACCheck -StorageAccountName $StorageAccountName -UserPrincipalName $UserPrincipalName -checkResult $checks["CheckDefaultShareRBACPermission"]
+                        #$DefaultSharePermission = "Not Configured. Please visit https://docs.microsoft.com/en-us/azure/storage/files/storage-files-identity-ad-ds-assign-permissions?tabs=azure-portal for more information if needed."
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4033]: Should we remove this comment?

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 5]
```diff
@@ -4019,17 +4016,26 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                    }
+                    Write-Verbose "CheckDefaultShareRBACPermission: $CheckDefaultShareRBACPermission"
+                    Write-Verbose "CheckDefaultShareRBACPermission - SUCCESS"
+                    $checks["CheckDefaultShareRBACPermission"].Result = "Passed"
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4037]: If `Debug-RBACCheck` is responsible for writing failures in the `CheckResult` object, then I suggest we also move the responsibility of writing successes into that function.

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 6]
```diff
@@ -4047,11 +4053,10 @@ function Debug-RBACCheck {
     param (
         [Parameter(Mandatory=$True, Position=0, HelpMessage="Storage account name")]
         [string]$StorageAccountName,
-        [Parameter(Mandatory=$True, Position=0, HelpMessage="User Principal name")]
+        [Parameter(Mandatory=$True, Position=1, HelpMessage="User Principal name")]
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4056]: Good catch!

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 7]
```diff
@@ -3755,7 +3755,7 @@ function Debug-AzStorageAccountEntraKerbAuth {
         [string]$Filter,
 
         [Parameter(Mandatory=$False, Position=3, HelpMessage="Optional parameter for filter 'CheckSidHasAadUser' and 'CheckUserFileAccess'. The user name to check.")]
-        [string]$UserName,
+        [string]$UserPrincipalName,
```
[the remark this answers is not in the document]
Comment (MainiSuruchi) [reply AzFilesHybrid/AzFilesHybrid.psm1:3758]: i did change this here. idk why it doesnt show here

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 8]
```diff
@@ -4006,11 +4007,102 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+    )
+    process {
+        try {
+            Request-ConnectMsGraph `
+                    -Scopes "User.Read.All", "GroupMember.Read.All" `
+                    -RequiredModules @("Microsoft.Graph.Users", "Microsoft.Graph.Groups", "Microsoft.Graph.Identity.DirectoryManagement")
+                    
+            $userOid = $(Get-MgUser -Filter "UserPrincipalName eq $UserPrincipalName" -Property Id).Id
+
+            $groupIds = Get-MgUserMemberOf -UserId $userOid | Select-Id
```
[the remark this answers is not in the document]
Comment (MaximeKjaer) [reply AzFilesHybrid/AzFilesHybrid.psm1:4065]: Do we need to address this?

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 9]
```diff
@@ -4006,11 +4004,118 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                    Write-Error "CheckRBAC - FAILED"
+                }
+                else 
+                {
+                    $DefaultSharePermission = $StorageAccountObject.AzureFilesIdentityBasedAuth.DefaultSharePermission
+                    
+                    if((!$DefaultSharePermission) -or ($DefaultSharePermission -eq 'None'))
+                    {
+                        Debug-RBACCheck -StorageAccountName $StorageAccountName -UserPrincipalName $UserPrincipalName -checkResult $checks["CheckRBAC"]
+                    }
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4033]: Let's add an `else` statement logging that access is granted via the default share permission

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 10]
```diff
@@ -4006,11 +4004,118 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                }
+            }
+
+            if ($listOfRoleNames.size -eq 0) {
+                $message = "User '$($user.UserPrincipalName)' is not assigned any SMB share-level permission to" `
+                        + " storage account '$StorageAccountName' in resource group '$ResourceGroupName'. Please" `
+                        + " configure proper share-level permission following the guidance at" `
+                        + " https://docs.microsoft.com/en-us/azure/storage/files/storage-files-identity-ad-ds-assign-permissions"
+                    Write-Error -Message $message -ErrorAction Stop
+            }
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4106]: Same here, let's ad an `else` that logs the `$listOfRoleNames` and explains how your RBAC is given.

Comment (MainiSuruchi) [reply AzFilesHybrid/AzFilesHybrid.psm1:4106]: Didnt understand this else part

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 11]
```diff
@@ -4214,6 +4251,84 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+        [Parameter(Mandatory=$True, Position=2, HelpMessage="Check result object")]
+        [CheckResult]$checkResult
+    )
+    process {
+        try {
+            Request-ConnectMsGraph `
+                    -Scopes "User.Read.All", "GroupMember.Read.All" `
+                    -RequiredModules @("Microsoft.Graph.Users", "Microsoft.Graph.Groups", "Microsoft.Graph.Identity.DirectoryManagement")
+                    
+            $userOid = $(Get-MgUser -Filter "UserPrincipalName eq $UserPrincipalName" -Property Id).Id
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4270]:
```suggestion
            $user = Get-MgUser -Filter "UserPrincipalName eq '$UserPrincipalName'" -Property Id,OnPremisesSecurityIdentifier
            
            if (!$user.OnPremisesSecurityIdentifier) {
                $checkResult.Result = "Failed"
                $checkResult.Issue = "User is a cloud-only user, cannot have RBAC access"
                Write-Error "CheckRBAC - FAILED"
                return
            }
            
            $groups = Get-MgUserMemberOfAsGroup -UserId $user.Id -Property DisplayName,Id,OnPremisesSecurityIdentifier
            
            $hybridGroups = $groups | Where-Object { $_.OnPremisesSecurityIdentifier } 
            $hybridGroupIds = $hybridGroups.Id
```

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 12]
```diff
@@ -4214,6 +4251,84 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                    {
+                        if ($assignment.ObjectId -eq $userOid) 
+                        {
+                            $listOfRoleNames.Add($roleName)
+                            break
+                        }
+                    }
+                    elseif ($assignment.ObjectType -eq "Group") 
+                    {
+                        if ($groupIds -contains $assignment.ObjectId)
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4298]:
```suggestion
                        if ($hybridGroupIds -contains $assignment.ObjectId) 

```

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 13]
```diff
@@ -4214,6 +4251,84 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                    Write-Error -Message $message -ErrorAction Stop
+            }
+            else {
+                Write-Host "You have access to the shares: Here is the list of Roles you have: "
+                foreach ($role in $listOfRoleNames.Keys) {
+                Write-Output $role
+            }
+
+
+        }
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4321]: Nit: spacing

File: AzFilesHybrid/AzFilesHybrid.psm1 (hunk 10 above)
Comment (MaximeKjaer) [reply AzFilesHybrid/AzFilesHybrid.psm1:4313]: Looks good now :)

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 14]
```diff
@@ -4060,6 +4058,45 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                else 
+                {
+                    $DefaultSharePermission = $StorageAccountObject.AzureFilesIdentityBasedAuth.DefaultSharePermission
+                    
+                    if((!$DefaultSharePermission) -or ($DefaultSharePermission -eq 'None'))
+                    {
+                        Debug-RBACCheck -StorageAccountName $StorageAccountName -UserPrincipalName $UserPrincipalName -checkResult $checks["CheckRBAC"]
+                    }
+                    else {
+                        Write-Host "Access is granted via the default share permission"
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4089]:
```suggestion
                        $checks["CheckRBAC"].Result = "Passed"
                        Write-Host "Access is granted via the default share permission"

```

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 15]
```diff
@@ -4214,6 +4251,99 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                Write-Error "CheckRBAC - FAILED"
+                return
+            }
+            
+            $groups = Get-MgUserMemberOfAsGroup -UserId $user.Id -Property DisplayName,Id,OnPremisesSecurityIdentifier
+            
+            $hybridGroups = $groups | Where-Object { $_.OnPremisesSecurityIdentifier } 
+            $hybridGroupIds = $hybridGroups.Id
+
+            $groupIds = (Get-MgUserMemberOf -UserId $userOid).Id
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4284]:
```suggestion
```

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 16]
```diff
@@ -4214,6 +4251,99 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+            $roleNames = @(
+                "Storage File Data SMB Share Reader",
+                "Storage File Data SMB Share Contributor",
+                "Storage File Data SMB Share Elevated Contributor"
+            )
+
+            $listOfRoleNames = @{}
+            foreach ($roleName in $roleNames) 
+            {
+                $scope = "/subscriptions/$subscription/resourceGroups/$resourceGroup/providers/Microsoft.Storage/storageAccounts/$storageAccount/fileServices/default/fileshares/$fileShare"
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4295]: Nitpick: bring this out of the loop

```suggestion
            $scope = "/subscriptions/$subscription/resourceGroups/$resourceGroup/providers/Microsoft.Storage/storageAccounts/$storageAccount/fileServices/default/fileshares/$fileShare"
            $listOfRoleNames = @{}
            foreach ($roleName in $roleNames) 
            {
```

File: AzFilesHybrid/AzFilesHybrid.psm1 [hunk 17]
```diff
@@ -4214,6 +4251,99 @@ function Debug-AzStorageAccountEntraKerbAuth {
...
+                }
+                else 
+                { 
+                    Write-Host "You have access to the shares: Here is the list of Roles you have: "
+                    foreach ($role in $listOfRoleNames.Keys) 
+                    {
+                    Write-Output $role
+                    }  
+
+
```
Comment (MaximeKjaer) [inline AzFilesHybrid/AzFilesHybrid.psm1:4335]:
```suggestion
                    $checkResult.Result = "Passed"
                    Write-Host "You have access to the shares: Here is the list of Roles you have: "
                    foreach ($role in $listOfRoleNames.Keys) 
                    {
                        Write-Output $role
                    }
```

**Pull request merged**